File: Dynamic_programming/효율적인화폐구성.py
n, m = map(int, input().split())
money = []
for _ in range(n):
    money.append(int(input()))
d = [10000] * 10001
d[0] = 0
 
for i in range(money[0], m + 1):
    for k in money:
        if d[i - k] != 10000:
            d[i] = min(d[i], d[i - k] + 1)
print(d[m]) if d[m] != 10000 else -1

File: Dynamic_programming/효율적인화폐구성.js
let inputValue = prompt()
  .split(" ")
  .map((v) => parseInt(v)); // "2 15"
let n = inputValue[0];
let m = inputValue[1];
let money = [];
for (let _ = 0; _ < n; _ += 1) {
  money.push(parseInt(prompt()));
}
let d = Array(10001).fill(10000);
d[0] = 0;
// money 3 5 7
for (let i = 0; i < n; i += 1) {
  for (let j = money[i]; j <= m; j += 1) {
    if (d[j - money[i]] !== 10000) {
      d[j] = Math.min(d[j], d[j - money[i]] + 1);
    }
    console.log(d[j]);
  }
}
console.log(d[m] !== 10000 ? d[m] : -1);
